### PR TITLE
Improve documentation of rewrite_strat/innermost and outermost

### LINKED ISSUE #13701: rewrite_strat strategies outermost and innermost are mis-named and mis-described in the documentation
#### Description of the problem

Consider
```coq
Require Import Coq.Classes.Morphisms Coq.Setoids.Setoid.

Goal (forall a b c d e f g h : nat, a + 0 = a) <-> (forall a : nat, a + 0 = a).
  rewrite_strat outermost <- plus_n_O.
```

Even though the left-hand-side `a + 0` is deeper in the term than the right-hand-side one, `outermost` still chooses the left one.  This matches the definition given at the bottom of https://coq.inria.fr/refman/addendum/generalized-rewriting.html#definitions of `outermost rewstrategy := fix ident. (choice rewstrategy (subterm ident))`, but does not match the "outermost first" description.  To me, "outermost first" suggests a left-to-right breadth-first search, not the first occurrence in a pre-order tree traversal.  Similarly, "innermost first" suggests to me a reversed breadth-first search (i.e., the last occurrence in a breadth-first search), not the first occurrence in a post-order tree traversal.

cc @mattam82 @jfehrle 

#### Coq Version

8.12

### PR DESCRIPTION
**Pull request by jfehrle**
Fixes #13701

### CODE REVIEW

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (gares) [inline doc/sphinx/addendum/generalized-rewriting.rst:736]: what is `s` here?

Comment (JasonGross) [reply doc/sphinx/addendum/generalized-rewriting.rst:736]:
```suggestion
   Rewrite using the strategy `@rewstrategy` in hypothesis `@ident` or the conclusion.
```

(maybe you need some sort of :something: before these?)

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (JasonGross) [inline doc/sphinx/addendum/generalized-rewriting.rst:740]:
```suggestion
      If the strategy failed to find any matches.
```

(There are other kinds of failures, like "No constraint can apply on a dependent ..." and type errors (which are generally bugs.)

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (JasonGross) [inline doc/sphinx/addendum/generalized-rewriting.rst:744]:
```suggestion
      If the strategy succeeded but made no progress.
```

This is not quite accurate.  Some strategies (`try` and `eval`, maybe some others) are allowed to succeed with no progress, and do not make `rewrite_strat` error.  Other strategies error if there is no progress.

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (JasonGross) [inline doc/sphinx/addendum/generalized-rewriting.rst:914]:
```suggestion
        rewrite_strat innermost Nat.add_comm.
```

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (JasonGross) [inline doc/sphinx/addendum/generalized-rewriting.rst:911]: Do you not also need `Require Import Coq.Setoids.Setoid.`?

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (JasonGross) [inline doc/sphinx/addendum/generalized-rewriting.rst:925]:
```suggestion
        rewrite_strat outermost Nat.add_comm.
```

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (JasonGross) [inline doc/sphinx/addendum/generalized-rewriting.rst:752]: Do you want to comment on the differences?  @mattam82 can you explain exactly what the differences are?

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (jfehrle) [reply doc/sphinx/addendum/generalized-rewriting.rst:736]: Yes, `:n:`

Comment (jfehrle) [reply doc/sphinx/addendum/generalized-rewriting.rst:736]: > what is `s` here?

Dunno, I only moved the text.  I will fix though.

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (jfehrle) [reply doc/sphinx/addendum/generalized-rewriting.rst:744]: This is describing a message which BTW only appears with typeclasses debug.  I think it's fine.

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (jfehrle) [reply doc/sphinx/addendum/generalized-rewriting.rst:911]: No.

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (jfehrle) [reply doc/sphinx/addendum/generalized-rewriting.rst:752]: No.  Let's not leverage responding to your initial comments into rewriting the whole thing.

**Comment by jfehrle** [comment]
Updated.

**Comment by Zimmi48** [comment]
@JasonGross Would you like to self-assign this PR? (Note that if we want it for 8.13.0, it should be merged soon.)

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (JasonGross) [reply doc/sphinx/addendum/generalized-rewriting.rst:744]: Ah, strange, but true

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (JasonGross) [reply doc/sphinx/addendum/generalized-rewriting.rst:752]: Ah, okay.  Another question here.  Why the current thing rather than
```suggestion
   The `setoid_rewrite c` tactic is basically equivalent to
   `rewrite_strat (outermost c)`.
```
or
```suggestion
   The :g:`setoid_rewrite c` tactic is basically equivalent to
   :g:`rewrite_strat (outermost c)`.
```
or
```suggestion
   The :n:`setoid_rewrite @one_term` tactic is basically equivalent to
   :n:`rewrite_strat (outermost @one_term)`.
```

**Review by JasonGross — APPROVED** [review]
LGTM modulo my question about ``` ``setoid_rewrite c`` ``` vs the other ways of writing this.  Does this need a review from @coq/doc-maintainers before I merge?

**Review by Zimmi48 — APPROVED** [review]
I've checked the form but not the content.

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (jfehrle) [reply doc/sphinx/addendum/generalized-rewriting.rst:752]: Both ``` ` ``` and ``` `` ``` give nonbold monospaced text,  ``` :n:` ``` enables regex graphics (as in the tactic definitions--boxes and hyperlinks to nonterminals with ``` @nonterminal ```).  ``` :g:` ``` seems to be the same as ``` ` ```, I haven't looked into it.

image

Unfortunately, `:n:` doesn't yet handle hyperlinks to tactics and commands, so I'll do it this way:

```
   :tacn:`setoid_rewrite` :n:`@one_term` is basically equivalent to
   :n:`rewrite_strat outermost @one_term`.
```

image

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (jfehrle) [inline doc/sphinx/addendum/generalized-rewriting.rst:846]: I moved these without giving them much thought.  In fact, only the one for try looks reasonable.  `fix` is not a rewstrategy.  There is a tactic with that name, but it takes 2 arguments (ident natural) and it begins a proof--that seems incorrect.  `try @ident` doesn't make sense, maybe they should be `try @rewstrategy`.
The ones with semicolons probably should be parenthesized so they aren't parsed as `tactic1 ; tactic2`.
If you give me the right expressions, I will add them.  Otherwise I will leave them as-is.

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (jfehrle) [reply doc/sphinx/addendum/generalized-rewriting.rst]: I couldn't see a difference in behavior with or without the `<-`, trying with:

```
Require Import Coq.Arith.Arith.
Set Printing Parentheses.
Goal forall a b c: nat, (a + b + c)*(a + b) = 0.
  rewrite_strat innermost Nat.add_comm.
```

If you can explain and it's simple, I'll add that, too.

**Comment by jfehrle** [comment]
Updated.  If you have answers to my 2 questions, I'll add those, otherwise I think it can be merged this way.  (The criterion is that we make things better, not that we fix everything.)

**Comment by gares** [comment]
last call for 8.13.0

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (JasonGross) [reply doc/sphinx/addendum/generalized-rewriting.rst:846]: The expressions are fine.  Confusingly enough, due to parsing rules, if you want to sequence a tactic after `rewrite_strat`, I think you need to do `(rewrite_strat foo); bar`.  It's much like `now` in that way.

The `fix` operator is not exposed, so these are pseudo-strategy expressions.  The `try @ident` is invoking `try` on the ident bound by `fix`, which denotes a strategy.

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (JasonGross) [inline doc/sphinx/addendum/generalized-rewriting.rst:837]: I think it's important to not drop this line (and the one after it), since the primitive fixpoint operator isn't actually exposed.

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (JasonGross) [reply doc/sphinx/addendum/generalized-rewriting.rst:846]: Oh, but you dropped the line mentioning the primitive fixpoint operator which isn't actually exposed to the user, which I think makes moving them make things worse.  I think you should either move them back or else leave a note that the definitions using `fix` are using a non-user-accessible primitive just to explain the meaning.

**Review by JasonGross — CHANGES_REQUESTED** [review]
I missed that you had dropped the line mentioning primitive fixpoint operator.  Sorry for not catching this earlier, but I think it's important to include.

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (jfehrle) [reply doc/sphinx/addendum/generalized-rewriting.rst:837]: Updated.  I reverted moving these descriptions.

The original text leaves much to be desired:
- It's not obvious that `fix` corresponds to "primitive fixpoint operator"
- No description of `fix`
- It's perverse to give equivalences using an operator that's not available to users.
- The description for `@one_term` only mentions lemmas, not fixpoint values
- `fix @ident.` probably should have a semicolon rather than a period (?)

If it's in the code, it appears to work and it's useful, would you care to file an issue asking to make it accessible and documented?

I don't see how `any @rewstrategy := fix @ident. try (@rewstrategy ; @ident)` would cause repetition.

Should `try @rewstrategy := choice @rewstrategy id` be `try @rewstrategy := fix @ident; choice @rewstrategy @ident`?  If so, I will fix that (and periods -> semicolons).

**Comment by jfehrle** [comment]
Jason, can we put this to bed soon?

**Comment by jfehrle** [comment]
@JasonGross Would you take another look at this and, if agreeable, merge this?  I think it will only take you a few minutes at most.  Thanks.

**Comment by jfehrle** [comment]
CI failures are not relevant for a doc-only PR.

**Comment by JasonGross** [comment]
Sorry for the delay and for dropping the ball on this

**Comment by JasonGross** [comment]
@coqbot: merge

**Review by JasonGross — APPROVED** [review]

**Comment by JasonGross** [comment]
@coqbot: merge now

File: doc/sphinx/addendum/generalized-rewriting.rst
Comment (JasonGross) [reply doc/sphinx/addendum/generalized-rewriting.rst:837]: @jfehrle 

> * It's not obvious that `fix` corresponds to "primitive fixpoint operator"

Indeed, though `fix` is Gallina's primitive fixpoint operator, so it's not much of a stretch to imagine a similar construct to Gallina `fix` for rewrite strategies

> - No description of `fix`

To run `fix <ident>. <body>`, replace all occurrences of `<ident>` in `<body>` with `fix <ident>. <body>` and then run the resulting code.

> - It's perverse to give equivalences using an operator that's not available to users.

I'd say unfortunate.  It is very informative to users like me, but it would be better if it were available.


> * The description for `@one_term` only mentions lemmas, not fixpoint values

Yes, this is equivalent to fixpoint values not being user accessible.

> * `fix @ident.` probably should have a semicolon rather than a period (?)

No, treat `fix` like a lambda.  Often in math λ arguments are followed by `.`, sometimes by `,`, sometimes by `=>` or `->` or `⇒` or `→` or `↦`, but I've never seen them followed by a `;`

> If it's in the code, it appears to work and it's useful, would you care to file an issue asking to make it accessible and documented?

https://github.com/coq/coq/issues/13702

> I don't see how `any @rewstrategy := fix @ident. try (@rewstrategy ; @ident)` would cause repetition.

Using "⇝" to mean "evaluates to", we have `any @rewstrategy` ⇝ `fix @ident. try (@rewstrategy ; @ident)` ⇝ `try (@rewstrategy ; (fix @ident. try (@rewstrategy ; @ident)))` ⇝ `try (@rewstrategy ; try (@rewstrategy ; (fix @ident. try (@rewstrategy ; @ident))))`.  There's the source of repetition.

> Should `try @rewstrategy := choice @rewstrategy id` be `try @rewstrategy := fix @ident; choice @rewstrategy @ident`?  If so, I will fix that (and periods -> semicolons).

No.  Your code gives an infinite loop on failure of `@rewstrategy`, because the fixpoint here is `choice @rewstrategy (choice @rewstrategy (choice @rewstrategy (choice @rewstrategy ...)))`, and we really want `choice @rewstrategy id` (i.e., act as the identity when `@rewstrategy` fails).

Comment (jfehrle) [reply doc/sphinx/addendum/generalized-rewriting.rst:837]: Thanks for explaining and the review.  My goal for the long term is to make the doc understandable to readers who only have a bachelor's degree in math or CS.  But lots to do to get there...

**Pull request merged**
